eliminate global static GCS_MAVLINK::send_home_all

peterbarker opened the pull request.

OXINARF commented:
LGTM, but I'll make two notes before merging:

- behavior changes as it will no longer only send home to active channels
- a new *chan* method wasn't needed, only making the current one be const (instead of the new one which is const and returns a const reference).

peterbarker commented:
@OXINARF a const function returning a reference must return a const reference, which is unfortunate if you would like to fiddle with the returned value....

I've changed the GCS_MAVLink patch to take into account the active channel mask.  That mask is going to die in future, replaced with a `.active` or `.active()` methodcall.  Similarly with the streaming channel mask.

amilcarlucas commented:
Sorry, I do not get the advantage of this PR. Can you elaborate Please ?

magicrub commented:
@peterbarker has had this task in the queue for a long time. We've been wanting to change the whole GCS messaging system to use a singleton instead of all the libraries and vehicles  reaching back and forth over the GCS_MAVLink:: const reference which is a bit ugly. He has a series of PRs lined up to totally remove the GCS const refs which were held off to make sure it doesn't conflict with the copter release

OXINARF commented:
> a const function returning a reference must return a const reference, which is unfortunate if you would like to fiddle with the returned value....

@peterbarker Ah you are right and it makes sense, you can't mark the method const if you give the caller the chance to change the internal state of the object. Learning every day! 🙂

@magicrub Why was this marked as DevCallTopic? This looks ready to go in, there is no change in behavior I can see.

OXINARF commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -1517,26 +1517,26 @@ void GCS_MAVLINK::send_home(const Location &home) const
+        }                                            \
+    } while (0);
+
+void GCS::send_home(const Location &home) const
```
Given we already have a GCS.cpp file should this go there? I think it would be better, easier to find.

peterbarker replied:
Yep!  Hadn't been created when I originally made this PR, I think.

Moved!

OXINARF commented:
Merged, thanks! I've changed the commit message of the last commit here by adding "*to GCS_Dummy*".
